File: 2025/02/main.py
from functools import cache

@cache
def usable_factors(n):
    return [i for i in range(1, n // 2 + 1) if n % i == 0]

def part1():
    part1 = 0
    for part in open("input.txt").read().split(","):
        low,high = map(int, part.split("-"))
        first_half = str(low)[:len(str(low))//2] or "0"
        current = int(2*first_half)
        while current <= high:
            if current >= low:
                part1 += current
            first_half = str(int(first_half)+1)
            current = int(2*first_half)
    return part1

def part2():
    part2 = set()
    for part in open("input.txt").read().split(","):
        low,high = map(int, part.split("-"))
        current = low
        while current <= high:
            curr_len = len(str(current))
            for factor_size in usable_factors(curr_len):
                pattern = str(current)[:factor_size]
                reps = curr_len // factor_size
                inner_current = int(pattern*reps)
                while inner_current <= high:
                    if inner_current >= low:
                        part2.add(inner_current)
                    pattern = str(int(pattern)+1)
                    inner_current = int(pattern*reps)
            current = int("1" + "0"*curr_len)
    return sum(part2)

if __name__ == "__main__":
    print("Part 1:",part1())
    print("Part 2:",part2())

File: 2025/02/main.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func usableFactors(n int) []int {
	var factors []int
	for i := 1; i <= n/2; i++ {
		if n%i == 0 {
			factors = append(factors, i)
		}
	}
	return factors
}

func parseRanges() [][2]int {
	data, _ := os.ReadFile("input.txt")
	parts := strings.Split(strings.TrimSpace(string(data)), ",")

	ranges := make([][2]int, len(parts))
	for i, part := range parts {
		fmt.Sscanf(part, "%d-%d", &ranges[i][0], &ranges[i][1])
	}
	return ranges
}

func part1() int {
	ranges := parseRanges()
	total := 0

	for _, r := range ranges {
		low, high := r[0], r[1]

		lowStr := strconv.Itoa(low)
		halfLen := len(lowStr) / 2
		pattern := lowStr[:halfLen]
		if pattern == "" {
			pattern = "0"
		}

		halfInt, _ := strconv.Atoi(pattern)
		current, _ := strconv.Atoi(pattern + pattern)

		for current <= high {
			if current >= low {
				total += current
			}
			halfInt++
			pattern = strconv.Itoa(halfInt)
			current, _ = strconv.Atoi(pattern + pattern)
		}
	}

	return total
}

func part2() int {
	ranges := parseRanges()
	invalidSet := make(map[int]bool)

	for _, r := range ranges {
		low, high := r[0], r[1]

		current := low
		for current <= high {
			currStr := strconv.Itoa(current)
			currLen := len(currStr)

			for _, factorSize := range usableFactors(currLen) {
				pattern := currStr[:factorSize]
				reps := currLen / factorSize

				patternInt, _ := strconv.Atoi(pattern)
				repeated, _ := strconv.Atoi(strings.Repeat(pattern, reps))

				for repeated <= high {
					if repeated >= low {
						invalidSet[repeated] = true
					}
					patternInt++
					pattern = strconv.Itoa(patternInt)
					repeated, _ = strconv.Atoi(strings.Repeat(pattern, reps))
				}
			}

			current = int(math.Pow10(currLen))
		}
	}

	total := 0
	for num := range invalidSet {
		total += num
	}

	return total
}

func main() {
	fmt.Println("Part 1:", part1())
	fmt.Println("Part 2:", part2())
}
